cpp/ql/src/semmle/code/cpp/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -90,10 +90,17 @@ predicate localAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeT
   exists(TaintFunction f, Call call, FunctionInput inModel, FunctionOutput outModel |
     call.getTarget() = f and
     inModel.isReturnValueDeref() and
-    outModel.isQualifierObject() and
-    f.hasTaintFlow(inModel, outModel) and
     nodeFrom.(DataFlow::PostUpdateNode).getPreUpdateNode().asExpr() = call and
-    nodeTo.asDefiningArgument() = call.getQualifier()
+    f.hasTaintFlow(inModel, outModel) and
+    (
+      outModel.isQualifierObject() and
+      nodeTo.asDefiningArgument() = call.getQualifier()
+      or
+      exists(int argOutIndex |
+        outModel.isParameterDeref(argOutIndex) and
+        nodeTo.asDefiningArgument() = call.getArgument(argOutIndex)
```

Comment:
Am I correct in thinking that this case is added to give taint in the following case:
```cpp
f(p); // taint from *f(p) to *p
```
assuming that we have a model for `f` that specifies `input.isReturnValueDeref() and output.isParameterDeref(0)`?

Reply:
Yes.  The motivating case is:
```
ss3 << "123" << source()
```
which is really
```
operator<<(operator<<(ss3, "123"), source())
           2          3            1
```
Taint flows from parameter to parameter through the outer call to `operator<<`, i.e. from 1 to 2.  The issue is that it has to flow from 2 to 3 in order to taint `ss3` as it should.  This is a 'reverse taint' flow, but in the past they've always been from the return value to the qualifier so that's all we supported - in this case it's flow from the return value to a parameter (dereference).